flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitReadOperator.java

```diff
@@ -0,0 +1,368 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.streaming.api.functions.source;
+
+import org.apache.flink.api.common.io.CheckpointableInputFormat;
+import org.apache.flink.api.common.io.FileInputFormat;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.core.fs.FileInputSplit;
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.runtime.state.AbstractStateBackend;
+import org.apache.flink.runtime.state.StreamStateHandle;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.operators.TimestampedCollector;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.tasks.StreamTaskState;
+import org.apache.flink.util.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * This is the operator that reads the splits received from {@link FileSplitMonitoringFunction}.
+ * This operator will receive just the split descriptors and then read and emit records. This may lead
+ * to backpressure. To avoid this, we will have another thread actually reading the splits and
+ * another forwarding the checkpoint barriers. The two should sync so that the checkpoints reflect the
+ * current state.
+ * */
+public class FileSplitReadOperator<OUT, S extends Serializable> extends AbstractStreamOperator<OUT>
+	implements OneInputStreamOperator<FileInputSplit, OUT> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(FileSplitReadOperator.class);
+
+	private static final FileInputSplit EOF = new FileInputSplit(-1, null, -1, -1, null);
+
+	private transient SplitReader<S, OUT> reader;
+	private transient TimestampedCollector<OUT> collector;
+
+	private Configuration configuration;
+	private FileInputFormat<OUT> format;
+	private TypeInformation<OUT> typeInfo;
```

Comment:
This is a very subtle thing but not all `TypeInformation` are `Serializable` and none of them should be. This is a problem that we introduced a while back.

The way to do it is to implement `OutputTypeConfigurable`, there the `TypeSerializer` can be created. In `open()` you should then ensure that you actually have a `TypeSerializer`.

And yes, I know that no-one can really know this without having encountered a serialization problem once ...  😅